### ISSUE TITLE
Region with word 'generated' disables analysis for entire file

### ISSUE THREAD

**Issue by nnpcYvIVl** [SonarSource/sonar-dotnet#2228]
### Description

A region with the word generated anywhere in the title causes sonarlint to skip analysis for the entire file.

```csharp
namespace SonarLint_Analysis_Failure
{
    public static class ShadowedMember
    {
        public static int MyProperty { get; set; }
        public class Initializer
        {
            public int MyProperty { get; set; } // S3218 should be thrown here
        }

        #region Requests generated from remote systems need to be tested here
        #endregion
    }
}
```

### Related information

* SonarC# Version: Latest
* Visual Studio Version: 15.9.4

**Comment by nnpcYvIVl** [comment]
@Evangelink Thanks for looking into it.

At the very least shouldn't the analysis only be disabled inside the region and not for the entire file? We have very large code files getting completely skipped by the analyzer because of these regions.

**Comment by Evangelink** [comment]
It should clearly! That's why I called the current mechanism a hack.

The problem is the time it would take to fix this implementation.

Do you have any idea of how many files are broken because of this behavior? It might help with changing the decision.

**Comment by nnpcYvIVl** [comment]
@nnpcYvIVl I don't think it's very many but since it's a silent failure I'm not sure. Members of our team were scratching our heads for weeks wondering some files were being analyzed but others weren't. Files that weren't being analyzed were known to have analysis results at one point so that made it more confusing.

Is it possible for another analyzer to flag when this analyzer is enabled? e.g. S9999 Analysis disabled for current file. This would make users aware of the problem and give them a chance to fix the file.

**Comment by pavel-mikula-sonarsource** [comment]
We've hit this problem ourselves #4048

@Evangelink, do you remember what was the affected WinForms project type or file? Was it manually written, or generated by some designer?

Thank you

**Comment by Evangelink** [comment]
I don't have all details fresh in mind and also it's something that was implemented before me but from what I recall we had some paying customers with old codebases from a time before generated code was put into partial classes. This generated code was raising lots of issues that the users didn't wanted to fix, my investigation concluded that when the hack was put in place there wasn't much way to allow this specific exclusion (roslyn and sonarqube APIs weren't advanced enough). 

The biggest problem was that we had only one unit test which wasn't really representative of the real cases. I would recommend to either:

- ask the support/consulting team to reach out to customers with old .net codebases to ask if they have cases of these regions to have some kind of estimate of the impact. It might be a bit tricky to have the answer.

- discuss with PM and support/consulting about this edge case and agree upon taking the risk to have a breaking change (will need to be introduced in a major version of the analyzer), knowing that you might have some complaints but at least you will have real life scenarios to work with. 

I have the feeling that if there are complaints that will most likely be because of all the new (true positive) files analyzed rather than false positives.

Note that when I did the investigation (short one), I could not find any easy (unexpensive) way to check for the region in the reporting helper.

**Comment by andrei-epure-sonarsource** [comment]
The logic in [GeneratedCodeRecognizer](https://github.com/SonarSource/sonar-dotnet/blob/1a6c17d9b9f615cd80b38305cd8a422a3442f206/analyzers/src/SonarAnalyzer.Common/Helpers/GeneratedCodeRecognizer.cs#L28) should stay the same. Only [CSharpGeneratedCodeRecognizer](https://github.com/SonarSource/sonar-dotnet/blob/1a6c17d9b9f615cd80b38305cd8a422a3442f206/analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpGeneratedCodeRecognizer.cs#L51) will be modified.

**Comment by andrei-epure-sonarsource** [comment]
(note for us: I didn't validate on peach, it wouldn't be very easy to do and our test suite is good enough for this scenario)

**Comment by duncanp-sonar** [comment]
@andrei-epure-sonarsource it's not obvious from the title or comments of this ticket that the detection of generated regions has been dropped completely. It would be useful to make it a bit more explicit e.g. by updating the title or adding some text in the original comment.

**Comment by andrei-epure-sonarsource** [comment]
Thanks @duncanp-sonar . I've updated the release notes, as well.
